### PR TITLE
[Cute, Bwd, Sm100] support varlen for fa4 bwd

### PR DESCRIPTION
**Pull request by endurehero**
hi @drisspg @tridao  , I support varlen for sm100 bwd. It has passed the unit tests on both MHA and GQA. Here is the test script I used. Please review this PR if you think it will help.

---------------------------------------------
```python
import torch
from functools import partial

from flash_attn.cute.interface import _flash_attn_fwd, _flash_attn_bwd


def get_diff(name, real, custom, is_print=True):
    real = real.reshape(-1)
    custom = custom.reshape(-1)
    max_index = torch.argmax(torch.abs(real - custom)).item()
    real_val = real[max_index].item()
    custom_val = custom[max_index].item()
    
    max_diff = abs(real_val - custom_val)
    if is_print:
        print(name + " max diff", max_diff, "index@", max_index, real_val, custom_val)
    return max_diff

def attention_ref(
    q: torch.Tensor,  # [total_query_len, num_q_heads, head_dim]
    k: torch.Tensor,  # [total_key_len, num_k_heads, head_dim]
    v: torch.Tensor,  # [total_key_len, num_k_heads, head_dim]
    cu_seqlens_q: torch.Tensor,
    cu_seqlens_k: torch.Tensor,
    max_seqlen_q: int,
    max_seqlen_k: int,
    sm_scale: float = None,
    causal: bool = True,
) -> torch.Tensor:
    total_query_len, num_q_heads, head_dim = q.shape
    total_key_len, num_k_heads, _ = k.shape
    num_share_q_heads = num_q_heads // num_k_heads
    batch_size = cu_seqlens_q.shape[0] - 1
    if sm_scale is None:
        sm_scale = 1.0 / math.sqrt(head_dim)
    # get mask
    mask = torch.zeros(
        (total_query_len, total_key_len), dtype=torch.bool, device=q.device
    )
    for b in range(batch_size):
        q_len, k_len = (
            cu_seqlens_q[b + 1] - cu_seqlens_q[b],
            cu_seqlens_k[b + 1] - cu_seqlens_k[b],
        )
        k_max_ids = (
            torch.arange(k_len, device=q.device)
        )
        q_ids = torch.arange(q_len, device=q.device)
        mask[
            cu_seqlens_q[b] : cu_seqlens_q[b + 1], cu_seqlens_k[b] : cu_seqlens_k[b + 1]
        ] = (q_ids[:, None] >= k_max_ids[None, :]) if causal else True
    # attention
    qk = (
        torch.einsum("qhd,khd->hqk", q, k.repeat_interleave(num_share_q_heads, 1))
        * sm_scale
    )

    
    qk = qk.masked_fill_(~mask[None, ...], -torch.inf)
    # query from beginning of the sequence can't attend to any compressed key
    qk = qk.softmax(dim=-1, dtype=torch.float32)
    qk = qk.nan_to_num(0)
    attn_output = torch.einsum(
        "hqk,khd->qhd", qk.to(v.dtype), v.repeat_interleave(num_share_q_heads, 1)
    )
    return attn_output

def test():
    torch.manual_seed(42)
    num_heads = 8
    num_heads_kv = 1
    head_dim = 128
    deterministic = False
    sm_margin = 0
    use_torch_profiler = False
    causal = True

    seqlens = torch.LongTensor([8192, 176, 1, 23481, 23]).int().cuda()
    cu_seqlens = torch.cat(
        [
            torch.zeros(1, dtype=torch.int32, device="cuda"),
            torch.cumsum(seqlens, dim=0),
        ],
        dim=0,
    ).to(torch.int32)
    batch = seqlens.shape[0]
    max_seqlen = seqlens.max().item()
    
    q = (
        torch.empty(cu_seqlens[-1], num_heads, head_dim, device="cuda")
        .uniform_(-1, 1)
        .to(torch.float16)
    )
    k = (
        torch.empty(cu_seqlens[-1], num_heads_kv, head_dim, device="cuda")
        .uniform_(-1, 1)
        .to(torch.float16)
    )
    v = (
        torch.empty(cu_seqlens[-1], num_heads_kv, head_dim, device="cuda")
        .uniform_(-1, 1)
        .to(torch.float16)
    )

    q_ref = q.clone().detach()
    k_ref = k.clone().detach()
    v_ref = v.clone().detach()
    q_ref.requires_grad_()
    k_ref.requires_grad_()
    v_ref.requires_grad_()
    q_ref.retain_grad()
    k_ref.retain_grad()
    v_ref.retain_grad()
    

    softmax_scale = head_dim**(-0.5)
    
    fwd_func = partial(_flash_attn_fwd, q=q, k=k, v=v, cu_seqlens_q=cu_seqlens, cu_seqlens_k=cu_seqlens, softmax_scale=softmax_scale, causal=causal, return_lse=True)
    out, lse = fwd_func()

    dout = torch.randn_like(q)
    dq = torch.empty_like(q)
    dk = torch.empty_like(k)
    dv = torch.empty_like(v)

    bwd_func = partial(_flash_attn_bwd, q=q, k=k, v=v, out=out, dout=dout, lse=lse, max_seqlen_k=max_seqlen ,cu_seqlens_q=cu_seqlens, cu_seqlens_k=cu_seqlens, dq=dq, dk=dk, dv=dv, softmax_scale=softmax_scale, causal=causal)
    bwd_func()

    
    out_ref = attention_ref(q_ref, k_ref, v_ref, cu_seqlens, cu_seqlens, max_seqlen, max_seqlen, softmax_scale, causal)
    get_diff("fwd ", out_ref, out)
    out_ref.backward(dout)
    get_diff("dq ", q_ref.grad, dq)
    get_diff("dk ", k_ref.grad, dk)
    get_diff("dv ", v_ref.grad, dv)

test()

```

### CODE REVIEW

**Comment by jayhshah** [comment]
We'll be merging in this branch [jshah/sm100-varlen-bwd](https://github.com/Dao-AILab/flash-attention/tree/jshah/sm100-varlen-bwd) to support varlen once I fix a deadlock issue with deterministic mode; I can also switch over to skipping extra dK/dV postprocess for MHA. I think it's essentially the same as your PR except that I use the varlen tile scheduler.

**Comment by endurehero** [comment]
get it
